rkt/pods.go

```diff
@@ -731,16 +733,85 @@ func (p *pod) getState() string {
 	return state
 }
 
+// Returns the pid of the child, or 0 if not ready
+func getChildPID(ppid int) (int, error) {
+	var pid int
+
+	// If possible, get the child in O(1). Fallback on O(n) when the kernel does not have
+	// either CONFIG_PROC_CHILDREN or CONFIG_CHECKPOINT_RESTORE
+	_, err := os.Stat("/proc/1/task/1/children")
+	if err == nil {
+		b, err := ioutil.ReadFile(fmt.Sprintf("/proc/%d/task/%d/children", ppid, ppid))
+		if err == nil {
+			children := strings.SplitN(string(b), " ", 2)
+			if len(children) == 2 && children[1] != "" {
+				return -1, fmt.Errorf("too many children of pid %d", ppid)
+			}
+			if _, err := fmt.Sscanf(children[0], "%d ", &pid); err == nil {
+				return pid, nil
+			}
+		}
+		return 0, nil
+	}
+
+	// Fallback on the slower method
+	fdir, err := os.Open(`/proc`)
+	if err != nil {
+		return -1, err
+	}
+	defer fdir.Close()
+
+	for {
+		fi, err := fdir.Readdir(1)
+		if err != nil {
+			return -1, err
+		}
+		if len(fi) == 0 {
```

Comment:
This test is unnecessary:

>     If n > 0, Readdir returns at most n FileInfo structures. In this case,
    if Readdir returns an empty slice, it will return a non-nil error
    explaining why. At the end of a directory, the error is io.EOF.